fw/http_parser.c

```diff
@@ -3632,14 +3632,36 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len,
 
 	__FSM_START(parser->state);
 
-	/* ----------------    Request Line    ---------------- */
+	/* - Skipping and stripping leading CRLFs - */
 
-	/* Parser internal initializers, must be called once per message. */
+	/* The parser accepts 1 optional CRLF or LF before the request line.
+	 * The parser stores the fact of presense of it for subsequent stripping.
+	 * The parser blocks the request if it contains additional CRLFs before the request line.
+	 */
 	__FSM_STATE(Req_0, hot) {
-		if (unlikely(IS_CRLF(c)))
-			__FSM_MOVE_nofixup(Req_0);
-		/* fall through */
+		if (unlikely(c == '\r')) {
+			__set_bit(TFW_HTTP_B_NEED_STRIP_LEADING_CR, req->flags);
+			__FSM_MOVE_nofixup(Req_0_Wait_LF);
+		}
+		if (unlikely(c == '\n')) {
+			__set_bit(TFW_HTTP_B_NEED_STRIP_LEADING_LF, req->flags);
+			__FSM_MOVE_nofixup(Req_0_CheckExtraLeadingCRLF);
```

Comment:
> `if (unlikely(c <= 0x20))`

I doubt in your vision. I saw the Req_Method code. At least 2 4-byte comarison and jump deeper if no match.
I believe, the `c <= 0x20` is a good way to offload Req_Method parser from garbaged messaged.
This is the only a comparison with character which is just loaded into register (see assembler code above).

Reply:
PS:
Probably I do not fully understand you because I still do not know how much the "state" cost?
Probably a good way is to move this check to the beginning of Req_Method?